File: windows-native/DataMover.Core/Services/MhlWriter.cs
using System.IO;
using System.Text;

namespace DataMover.Core.Services;

/// <summary>
/// [2026-09-03] Port 1:1 al MHLWriter.swift (Mac) — generator de fisier MHL
/// (Media Hash List, versiunea 1.1), standardul prin care un ofloader de
/// platou preda datele catre post-productie.
///
/// DE CE: pana acum DataMover producea CSV + PDF, adica rapoarte pe care le
/// citeste un OM. Un MHL e acelasi lucru, dar citit de MASINA: Silverstack,
/// YoYotta, ShotPut Pro, Resolve si orice casa de post pot re-verifica
/// automat, luni mai tarziu, ca fiecare fisier de pe LTO/NAS e bit-identic
/// cu ce a iesit din camera in ziua filmarii.
///
/// MEMORIE (Regula 21): intrarile NU se acumuleaza in RAM — fiecare `hash`
/// se scrie imediat intr-un fisier temporar `.part`, iar la Close() se
/// compune fisierul final (antet + corp). Corpul nu poate fi scris direct
/// in fisierul final pentru ca `creatorinfo` sta obligatoriu PRIMUL in
/// schema si contine `finishdate`, cunoscut abia la sfarsit.
/// </summary>
public sealed class MhlWriter : IDisposable
{
    /// Algoritmii acceptati de schema MHL 1.1. SHA-256/SHA-512 NU fac parte
    /// din standard — cu ele, verificarea si rapoartele CSV/PDF/HTML raman
    /// complete, doar MHL-ul nu se genereaza.
    public static string? ElementFor(VerificationModel model) => model switch
    {
        VerificationModel.Md5 => "md5",
        VerificationModel.Sha1 => "sha1",
        VerificationModel.XxHash64 => "xxhash64be",
        _ => null,
    };

    public static bool IsSupported(VerificationModel model) => ElementFor(model) != null;

    private readonly string _finalPath;
    private readonly string _partPath;
    private readonly string _hashElement;
    private readonly string _toolName;
    private readonly DateTime _startedAt;
    private StreamWriter? _part;

    public int EntryCount { get; private set; }

    private MhlWriter(string path, string element, string toolName, DateTime startedAt)
    {
        _finalPath = path;
        _partPath = path + ".part";
        _hashElement = element;
        _toolName = toolName;
        _startedAt = startedAt;
        _part = new StreamWriter(_partPath, append: false, Encoding.UTF8);
    }

    /// null daca algoritmul nu e in standardul MHL sau fisierul nu poate fi
    /// deschis — apelantul continua transferul normal, fara MHL.
    public static MhlWriter? TryCreate(string path, VerificationModel model, string toolName, DateTime startedAt)
    {
        var element = ElementFor(model);
        if (element == null) return null;
        try { return new MhlWriter(path, element, toolName, startedAt); }
        catch { return null; }
    }

    /// Un fisier verificat cu succes. Se apeleaza DOAR pentru fisierele cu
    /// status OK/SARIT — un MHL nu are voie sa contina un fisier care n-a
    /// trecut verificarea, altfel ar certifica date corupte.
    public void Add(string relPath, long size, DateTime? modified, string hashHex, DateTime hashedAt)
    {
        if (_part == null || string.IsNullOrEmpty(hashHex)) return;
        try
        {
            // Caile din MHL folosesc mereu `/`, indiferent de platforma —
            // altfel un MHL scris pe Windows n-ar putea fi verificat pe Mac.
            var normalized = relPath.Replace('\\', '/');
            _part.WriteLine("  <hash>");
            _part.WriteLine($"    <file>{Escape(normalized)}</file>");
            _part.WriteLine($"    <size>{size}</size>");
            if (modified.HasValue)
                _part.WriteLine($"    <lastmodificationdate>{Iso(modified.Value)}</lastmodificationdate>");
            _part.WriteLine($"    <{_hashElement}>{hashHex}</{_hashElement}>");
            _part.WriteLine($"    <hashdate>{Iso(hashedAt)}</hashdate>");
            _part.WriteLine("  </hash>");
            EntryCount++;
        }
        catch { /* o intrare pierduta nu opreste transferul */ }
    }

    /// Scrie fisierul MHL final. Intoarce calea lui, sau null daca n-a
    /// existat nicio intrare valida (nu lasam pe disc un MHL gol, care ar
    /// parea o certificare a zero fisiere).
    public string? Close(DateTime finishedAt)
    {
        try { _part?.Flush(); _part?.Dispose(); } catch { /* ignora */ }
        _part = null;
        try
        {
            if (EntryCount == 0) return null;
            var header = new StringBuilder();
            header.AppendLine("<?xml version=\"1.0\" encoding=\"UTF-8\"?>");
            header.AppendLine("<hashlist version=\"1.1\">");
            header.AppendLine("  <creatorinfo>");
            header.AppendLine($"    <name>{Escape(Environment.UserName)}</name>");
            header.AppendLine($"    <username>{Escape(Environment.UserName)}</username>");
            header.AppendLine($"    <hostname>{Escape(Environment.MachineName)}</hostname>");
            header.AppendLine($"    <tool>{Escape(_toolName)}</tool>");
            header.AppendLine($"    <startdate>{Iso(_startedAt)}</startdate>");
            header.AppendLine($"    <finishdate>{Iso(finishedAt)}</finishdate>");
            header.AppendLine("  </creatorinfo>");

            using (var output = new StreamWriter(_finalPath, append: false, Encoding.UTF8))
            {
                output.Write(header.ToString());
                using (var input = new StreamReader(_partPath, Encoding.UTF8))
                {
                    var buffer = new char[64 * 1024];
                    int read;
                    while ((read = input.Read(buffer, 0, buffer.Length)) > 0)
                        output.Write(buffer, 0, read);
                }
                output.WriteLine("</hashlist>");
            }
            return _finalPath;
        }
        catch { return null; }
        finally
        {
            try { if (File.Exists(_partPath)) File.Delete(_partPath); } catch { /* ignora */ }
        }
    }

    public void Dispose()
    {
        try { _part?.Dispose(); } catch { /* ignora */ }
        _part = null;
    }

    private static string Iso(DateTime value) => value.ToString("yyyy-MM-ddTHH:mm:sszzz");

    private static string Escape(string s) => s
        .Replace("&", "&amp;").Replace("<", "&lt;").Replace(">", "&gt;")
        .Replace("\"", "&quot;").Replace("'", "&apos;");
}

File: mac-native/Sources/DataMoverMac/MHLWriter.swift
import Foundation

/// [2026-09-03] Generator de fisier MHL (Media Hash List, versiunea 1.1) —
/// standardul de facto prin care un ofloader de platou preda datele catre
/// post-productie.
///
/// DE CE: pana acum DataMover producea CSV + PDF, adica rapoarte pe care le
/// citeste un OM. Un MHL e acelasi lucru, dar citit de MASINA: Silverstack,
/// YoYotta, ShotPut Pro, Resolve si orice casa de post pot re-verifica
/// automat, luni mai tarziu, ca fiecare fisier de pe LTO/NAS e bit-identic
/// cu ce a iesit din camera in ziua filmarii. Fara MHL, un card descarcat
/// cu DataMover nu poate intra intr-un flux profesional — cu el, poate.
///
/// Fisierul se scrie LANGA datele copiate (in radacina folderului de
/// destinatie) si contine cai RELATIVE la propria pozitie, exact ca cele
/// scrise de ShotPut Pro — altfel mutarea folderului pe alt disc ar
/// invalida tot.
///
/// MEMORIE (Regula 21): intrarile NU se acumuleaza in RAM. Fiecare `<hash>`
/// se scrie imediat intr-un fisier temporar `.part`; la `close()` se
/// compune fisierul final (antet + corp), citind `.part`-ul in bucati.
/// Motivul pentru care corpul nu poate fi scris direct in fisierul final:
/// `<creatorinfo>` sta obligatoriu PRIMUL in schema MHL si contine
/// `<finishdate>`, pe care il stim abia la sfarsit.
final class MHLWriter {
    /// Algoritmii pe care ii accepta schema MHL 1.1. SHA-256/SHA-512 NU
    /// fac parte din standard — daca userul alege unul dintre ele,
    /// verificarea si rapoartele CSV/PDF raman complete, doar MHL-ul nu se
    /// genereaza (vezi `element(for:)` -> nil).
    static func element(for model: VerificationModel) -> String? {
        switch model {
        case .md5: return "md5"
        case .sha1: return "sha1"
        case .xxhash64: return "xxhash64be"
        case .sha256, .sha512, .sizeOnly: return nil
        }
    }

    static func isSupported(_ model: VerificationModel) -> Bool { element(for: model) != nil }

    private let finalPath: String
    private let partPath: String
    private let hashElement: String
    private let toolName: String
    private let startedAt: Date
    private var partHandle: FileHandle?
    private(set) var entryCount = 0

    private static let iso: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    init?(path: String, model: VerificationModel, toolName: String, startedAt: Date) {
        guard let element = Self.element(for: model) else { return nil }
        self.finalPath = path
        self.partPath = path + ".part"
        self.hashElement = element
        self.toolName = toolName
        self.startedAt = startedAt
        FileManager.default.createFile(atPath: partPath, contents: nil)
        guard let handle = FileHandle(forWritingAtPath: partPath) else { return nil }
        self.partHandle = handle
    }

    /// Un fisier verificat cu succes. Se apeleaza DOAR pentru fisierele cu
    /// status OK/SARIT — un MHL nu are voie sa contina un fisier care n-a
    /// trecut verificarea, altfel ar certifica date corupte.
    func add(relPath: String, size: Int64, modificationDate: Date?, hashHex: String, hashedAt: Date) {
        guard let handle = partHandle, !hashHex.isEmpty else { return }
        var xml = "  <hash>\n"
        xml += "    <file>\(Self.escape(relPath))</file>\n"
        xml += "    <size>\(size)</size>\n"
        if let mod = modificationDate {
            xml += "    <lastmodificationdate>\(Self.iso.string(from: mod))</lastmodificationdate>\n"
        }
        xml += "    <\(hashElement)>\(hashHex)</\(hashElement)>\n"
        xml += "    <hashdate>\(Self.iso.string(from: hashedAt))</hashdate>\n"
        xml += "  </hash>\n"
        autoreleasepool {
            try? handle.write(contentsOf: Data(xml.utf8))
        }
        entryCount += 1
    }

    /// Scrie fisierul MHL final. Intoarce calea lui, sau nil daca n-a
    /// existat nicio intrare valida (nu lasam pe disc un MHL gol, care ar
    /// parea o certificare a zero fisiere).
    @discardableResult
    func close(finishedAt: Date) -> String? {
        try? partHandle?.close()
        partHandle = nil
        defer { try? FileManager.default.removeItem(atPath: partPath) }
        guard entryCount > 0 else { return nil }

        var header = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
        header += "<hashlist version=\"1.1\">\n"
        header += "  <creatorinfo>\n"
        header += "    <name>\(Self.escape(NSFullUserName()))</name>\n"
        header += "    <username>\(Self.escape(NSUserName()))</username>\n"
        header += "    <hostname>\(Self.escape(Host.current().localizedName ?? ProcessInfo.processInfo.hostName))</hostname>\n"
        header += "    <tool>\(Self.escape(toolName))</tool>\n"
        header += "    <startdate>\(Self.iso.string(from: startedAt))</startdate>\n"
        header += "    <finishdate>\(Self.iso.string(from: finishedAt))</finishdate>\n"
        header += "  </creatorinfo>\n"

        FileManager.default.createFile(atPath: finalPath, contents: nil)
        guard let out = FileHandle(forWritingAtPath: finalPath),
              let part = FileHandle(forReadingAtPath: partPath) else { return nil }
        defer { try? out.close(); try? part.close() }
        do {
            try out.write(contentsOf: Data(header.utf8))
            while true {
                var stop = false
                try autoreleasepool {
                    guard let chunk = try part.read(upToCount: 256 * 1024), !chunk.isEmpty else {
                        stop = true
                        return
                    }
                    try out.write(contentsOf: chunk)
                }
                if stop { break }
            }
            try out.write(contentsOf: Data("</hashlist>\n".utf8))
        } catch {
            return nil
        }
        return finalPath
    }

    private static func escape(_ s: String) -> String {
        s.replacingOccurrences(of: "&", with: "&amp;")
         .replacingOccurrences(of: "<", with: "&lt;")
         .replacingOccurrences(of: ">", with: "&gt;")
         .replacingOccurrences(of: "\"", with: "&quot;")
         .replacingOccurrences(of: "'", with: "&apos;")
    }
}
